polars/polars-core/src/chunked_array/ndarray.rs

```diff
@@ -131,18 +141,30 @@ impl DataFrame {
                     let ca = s.unpack::<N>()?;
                     let vals = ca.cont_slice().unwrap();
 
+                    // Depending on the desired order, we add items to the buffer.
                     // Safety:
-                    // we get parallel access to the vector
-                    // but we make sure that we don't get aliased access by offsetting the column indices + length
-                    unsafe {
-                        let offset_ptr = (ptr as *mut N::Native).add(col_idx * height);
-                        // Safety:
-                        // this is uninitialized memory, so we must never read from this data
-                        // copy_from_slice does not read
-                        let buf = std::slice::from_raw_parts_mut(offset_ptr, height);
-                        buf.copy_from_slice(vals)
+                    // We get parallel access to the vector by offsetting index access accordingly.
+                    // For C-order, we only operate on every num-col-th element, starting from the
+                    // column index. For Fortran-order we only operate on n contiguous elements,
+                    // offset by n * the column index.
+                    match ordering {
+                        IndexOrder::C => unsafe {
```

Comment:
Can we do some benchmarking for the C ordering one? We could be false sharing here. Can we add parallel and a non-parallel implementation on different dataset sizes?

Reply:
Yeah, I had the same thought, will do!

Reply:
Since I didn't find an obvious place in the repository: where do you usually put benchmarks; are they run as part of the CI?

Reply:
They should be run locally. We don't include them in the repo.

Reply:
Ok, so benchmarks do not show any (obvious) sign of false sharing. So far, I only benchmarked the current code (i.e. the parallel implementation, benchmark code can be found below) and got the following results:

- `small_df` (1k rows, 10 cols), dtype `f32`
- `large_df` (1M rows, 500 cols), dtype `f32`

|            | Fortran-Ordering | C-Ordering |
| ---------- | :--------------: | :--------: |
| `small_df` | 615.56 µs        | 608.47 µs  |

<details><summary><b>Benchmark Code</b></summary>

```rust
use criterion::{criterion_group, criterion_main, Criterion};
use polars_core::prelude::*;

fn small_df(c: &mut Criterion) {
    let mut group = c.benchmark_group("small_df");
    let col_count = 1000;
    let row_count = 10;
    let series = (0..col_count)
        .into_iter()
        .map(|v| Series::new(&format!("{}", v), vec![v as f32; row_count]))
        .collect();
    let df = DataFrame::new(series).unwrap();

    group.bench_function("to_ndarray_fortran", |b| {
        b.iter(|| df.to_ndarray::<Float32Type>(IndexOrder::Fortran))
    });
    group.bench_function("to_ndarray_c", |b| {
        b.iter(|| df.to_ndarray::<Float32Type>(IndexOrder::C))
    });
}

fn large_df(c: &mut Criterion) {
    let mut group = c.benchmark_group("large_df");
    let col_count = 1_000_000;
    let row_count = 500;
    let series = (0..col_count)
        .into_iter()
        .map(|v| Series::new(&format!("{}", v), vec![v as f32; row_count]))
        .collect();
    let df = DataFrame::new(series).unwrap();

    group.bench_function("to_ndarray_fortran", |b| {
        b.iter(|| df.to_ndarray::<Float32Type>(IndexOrder::Fortran))
    });
    group.bench_function("to_ndarray_c", |b| {
        b.iter(|| df.to_ndarray::<Float32Type>(IndexOrder::C))
    });
}

criterion_group!(benches, small_df, large_df);
criterion_main!(benches);
```

</details>

Reply:
Ok, I benchmarked the current code both with the parallel implementation as in this PR and a sequential implementation with data frames of different size (specified as `[nrow, ncol]`):

#### Parallel Implementation

| Size               | Fortran-Ordering | C-Ordering |
| :----------------: | :--------------: | :--------: |
| `[1_000, 10]`      | 79.2 µs          | 83.3 µs    |
| `[1_000_000, 500]` | 395.9 ms         | 1.68 s     |
| `[10_000, 50_000]` | 460.2 ms         | 697.4 ms   |
| `[100_000_000, 5]` | 535.0 ms         | 737.8 ms   |

#### Sequential Implementation

| Size               | Fortran-Ordering | C-Ordering |
| :----------------: | :--------------: | :--------: |
| `[1_000, 10]`      | 53.7 µs          | 58.4 µs    |
| `[1_000_000, 500]` | 542.5 ms         | 3.66 s     |
| `[10_000, 50_000]` | 657.8 ms         | 2.14 s     |
| `[100_000_000, 5]` | 635.6 ms         | 969.8 ms   |

All tests were run on an Apple M1 Pro (2 efficiency cores, 8 high-performance cores, 32 GB memory).

<details><summary><b>Benchmark Code</b></summary>

```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion};
use polars_core::prelude::*;

fn build_df(row_count: usize, col_count: usize) -> DataFrame {
    let series = (0..col_count)
        .into_iter()
        .map(|v| Series::new(&format!("{}", v), vec![v as f64; row_count]))
        .collect();
    DataFrame::new(series).unwrap()
}

fn benchmark(c: &mut Criterion) {
    let shapes = [
        (1_000usize, 10usize),
        (1_000_000, 500),
        (10_000, 50_000),
        (100_000_000, 5),
    ];
    for shape in shapes.into_iter() {
        let mut group = c.benchmark_group(format!("nrows={}, ncols={}", shape.0, shape.1));
        let df = build_df(shape.0, shape.1);
        for order in [IndexOrder::Fortran, IndexOrder::C].into_iter() {
            group.bench_with_input(format!("{:?}", order), &df, |b, df| {
                b.iter(|| black_box(df.to_ndarray::<Float32Type>(order)));
            });
        }
    }
}

criterion_group!(benches, benchmark);
criterion_main!(benches);
```

</details>

## Conclusion

IMO, there is nothing significant to gain by using a sequential implementation in some cases. The 4x increased runtime isn't great, of course, but inevitable in some cases to keep memory consumption low. Also, compared to a `numpy.reshape`, the performance difference is negligible.

Would you have any other idea to make things faster?

Reply:
Could you also run a `10 * 100_000` shape to see how it performs in the other stride direction?

Reply:
Pretty much indistinguishable performance, parallel implementation yields 38.07 ms for Fortran-order and 38.09 ms for C-order.

Reply:
Alright. Thanks for the benchmark. There will be much more cache trashing, so the the 4x seems expected. The code looks great, so I am merging this in. Thank you!